### PR TITLE
feat(samples): Twin configurations sample

### PR DESCRIPTION
**Pull request by bikamani**

### CODE REVIEW

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            IoTHubServiceClient = client;
+        }
+
+        public async Task RunSampleAsync()
+        {
+            string testConfigurationId = $"configuration{Random.Next(MaxRandomValue)}";
+            TwinConfiguration twinConfiguration = CreateSampleConfig(testConfigurationId);
+            
+            // Create a Twin Configuration.
+            await CreateConfiguratioAsync(twinConfiguration);
```
Comment (bikamani) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:30]: currently adding sample for CRUD, we will modify and add edge config and test queries when we have implementation for the same.

Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:30]: Configuratio_n_

Comment (drwill-ms) [reply sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:30]: Same misspelling below too.

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        public async Task<TwinConfiguration> CreateConfiguratioAsync(TwinConfiguration twinConfiguration)
+        {
+            SampleLogger.PrintHeader("CREATE TWIN CONFIGURATION");
+            TwinConfiguration createdConfig;
+
+            try
+            {
+                // Create a twin configuration
+                #region Snippet:IotHubCreateConfiguration
+                Response<TwinConfiguration> createResponse =
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:55]: Prefer this code layout:

```
 Response<TwinConfiguration> createResponse = await IoTHubServiceClient.Configurations
    .CreateOrUpdateConfigurationAsync(twinConfiguration)
    .ConfigureAwait(false);
```

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        {
+            SampleLogger.PrintHeader("UPDATE A CONFIGURATION");
+
+            try
+            {
+                #region Snippet:IotHubUpdateConfiguration
+
+                twinConfiguration.Priority = Random.Next(MaxRandomValue);
+                Console.WriteLine($"Updating twin configuration with Id: '{twinConfiguration.Id}''s priority to: '{twinConfiguration.Priority}'");
+                Response < TwinConfiguration > response = await IoTHubServiceClient.Configurations.CreateOrUpdateConfigurationAsync(twinConfiguration, IfMatchPrecondition.UnconditionalIfMatch).ConfigureAwait(false);
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:121]: There are extra spaces in the return type here

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        /// Deletes a twin configuration.
+        /// </summary>
+        /// <param name="twinConfiguration">Twin Configuration to deletec.</param>
+        public async Task DeleteConfiguratioAsync(TwinConfiguration twinConfiguration)
+        {
+            SampleLogger.PrintHeader("DELETE A CONFIGURATION");
+
+            try
+            {
+
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:150]: remove this extra newline

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                await DeleteConfiguratioAsync(twinConfiguration);
+                SampleLogger.FatalError($"Failed to update a twin configuration due to:\n{ex}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Deletes a twin configuration.
+        /// </summary>
+        /// <param name="twinConfiguration">Twin Configuration to deletec.</param>
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:143]: misspelled delete

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            }
+            catch (Exception ex)
+            {
+                SampleLogger.FatalError($"Failed to delete twin configuration due to:\n{ex}");
+            }
+        }
+
+        private TwinConfiguration CreateSampleConfig(string testConfigurationId)
+        {
+            var twinConfiguration =
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:169]: Preferred code layout:

```
var twinConfiguration = new TwinConfiguration
{
    Id = testConfigurationId,
};
```

In some cases when there is just one property, we also do:

```
var twinConfiguration = new TwinConfiguration { Id = testConfigurationId };
```

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                new TwinConfiguration
+                {
+                    Id = testConfigurationId
+                };
+
+            // Labels are optional but adding here due to null check failure in deserialization
+            // Also note that we are not setting Host Platform value from Environment since that'll fail in our build pipeline
+            twinConfiguration.Labels.Add("HostPlatform", "SomeValue");
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:177]: You should be able to add these into the initializer list:

```
var twinConfiguration = new TwinConfiguration
{
    Id = testConfigurationId,
    Labels =
    {
        { "HostPlatform", "SomeValue" },
    },
    Content = new ConfigurationContent
    {
        { "properties.desired.deviceContent_key", $"deviceContent_value-{twinConfiguration.Id}",
    },
};
```

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/Program.cs [hunk 8]
```diff
@@ -62,6 +62,9 @@ public static async Task Main(string[] args)
             var statisticsSample = new StatisticsSamples(hubClient);
             await statisticsSample.RunSampleAsync();
 
+            var configurationsSample = new ConfigurationSamples(hubClient);
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/Program.cs:65]: `configurationsSample` (plural) or `configurationSample`?

**Review by drwill-ms — APPROVED** [review]

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/Program.cs (hunk 8 above)
Comment (bikamani) [reply sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/Program.cs:65]: yeah agree, I saw other samples also with plural though :)

File: sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs (hunk 1 above)
Comment (bikamani) [reply sdk/iot/Azure.Iot.Hub.Service/samples/IotHubClientSamples/ConfigurationSamples.cs:30]: 👍

**Pull request merged**